File: backend/LocalMind.Api/Services/Chat/ChatService.cs
using System.Text;
using LocalMind.Api.DTOs.Documents;
using LocalMind.Api.Services.Ai;
using LocalMind.Api.Services.Rag;
using LocalMind.Api.Services.Tools;

namespace LocalMind.Api.Services.Chat;

public class ChatService : IChatService
{
    private readonly IOllamaService _ollamaService;
    private readonly IRagService _ragService;
    private readonly IToolIntentDetector _toolIntentDetector;
    private readonly IAiToolService _aiToolService;

    public ChatService(
        IOllamaService ollamaService,
        IRagService ragService,
        IToolIntentDetector toolIntentDetector,
        IAiToolService aiToolService)
    {
        _ollamaService = ollamaService;
        _ragService = ragService;
        _toolIntentDetector = toolIntentDetector;
        _aiToolService = aiToolService;
    }

    public async Task<ChatResult> GenerateResponseAsync(
        int userId,
        string message,
        CancellationToken cancellationToken = default)
    {
        var toolIntent = _toolIntentDetector.Detect(message);
        var isDocumentQuestion = _toolIntentDetector.IsDocumentQuestion(message);

        if (toolIntent is not ToolIntent.None && !isDocumentQuestion)
        {
            var toolResult = await _aiToolService.TryExecuteAsync(
                toolIntent,
                message,
                cancellationToken);

            if (toolResult is not null)
            {
                return new ChatResult
                {
                    Response = toolResult.Response,
                    UsedTool = true,
                    ToolName = toolResult.ToolName,
                    Route = "tool"
                };
            }
        }

        if (!isDocumentQuestion)
        {
            return new ChatResult
            {
                Response = await _ollamaService.SendMessageAsync(
                    message,
                    cancellationToken),
                UsedRag = false,
                Route = "chat"
            };
        }

        var ragResult = await _ragService.SearchAsync(
            userId,
            message,
            cancellationToken);

        if (!ragResult.HasContext)
        {
            return new ChatResult
            {
                Response = await _ollamaService.SendMessageAsync(
                    message,
                    cancellationToken),
                UsedRag = false,
                Route = "chat"
            };
        }

        var contextBuilder = new StringBuilder();

        foreach (var match in ragResult.Matches)
        {
            contextBuilder.AppendLine(
                $"Fuente: {match.FileName} | chunk {match.ChunkIndex} | relevancia {match.Score:P1}");

            contextBuilder.AppendLine(match.Content);
            contextBuilder.AppendLine("---");
        }

        const string systemPrompt =
            "Respondé siempre en español. Usá únicamente el contexto de documentos provisto cuando sea relevante. " +
            "Si el contexto no alcanza para responder, decilo claramente. Incluí una sección final llamada 'Fuentes' con el nombre del archivo y chunk usado.";

        var userPrompt =
            $"Contexto de documentos:\n{contextBuilder}\n\nPregunta del usuario:\n{message}";

        var response = await _ollamaService.SendMessageAsync(
            systemPrompt,
            userPrompt,
            cancellationToken);

        return new ChatResult
        {
            Response = response,
            UsedRag = true,
            Route = "rag",
            ChunksUsed = ragResult.Matches.Count,
            Sources = ragResult.Matches.Select(match => new RagSourceResponse
            {
                DocumentId = match.DocumentId,
                FileName = match.FileName,
                ChunkIndex = match.ChunkIndex,
                Score = Math.Round(match.Score, 4),
                Preview = match.Content.Length > 220
                    ? $"{match.Content[..220]}..."
                    : match.Content
            }).ToList()
        };
    }
}

File: frontend/src/features/chat/chatService.ts
import { api } from "../../services/api";

export interface ConversationResponse {
  id: number;
  title: string;
  messages: ChatMessage[];
}

export const getConversation = async (
  conversationId: number
): Promise<ConversationResponse> => {
  const response = await api.get<ConversationResponse>(
    `/chat/history/${conversationId}`
  );

  return response.data;
};
export interface RagSource {
  documentId: number;
  fileName: string;
  chunkIndex: number;
  score: number;
  preview: string;
}

export interface ChatMessage {
  role: "user" | "assistant";
  content: string;
  createdAt?: string;
    sources?: RagSource[];
  usedTool?: boolean;
  toolName?: string | null;
  route?: string;
}

export interface SendMessageRequest {
  conversationId: number | null;
  message: string;
}

export interface SendMessageResponse {
  conversationId: number;
  response: string;
    usedRag: boolean;
  usedTool: boolean;
  toolName: string | null;
  route: string;
  chunksUsed: number;
  sources: RagSource[];
}

export interface ConversationHistoryItem {
  id: number;
  title: string;
  createdAt: string;
}

export const sendMessage = async (
  data: SendMessageRequest
): Promise<SendMessageResponse> => {
  const response = await api.post<SendMessageResponse>("/chat/send", data);
  return response.data;
};

export const getHistory = async (): Promise<ConversationHistoryItem[]> => {
  const response = await api.get<ConversationHistoryItem[]>("/chat/history");
  return response.data;
};